### PR TITLE
feat: support for broadcast notifications

### PR DESCRIPTION
**Pull request by drodil**
## Hey, I just made a Pull Request!

the broadcast notifications are returned with the same queries as user notifications with only few exceptions in the store

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/notifications-backend/migrations/20240313_broadcast.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+exports.up = async function up(knex) {
+  await knex.schema.createTable('broadcast', table => {
+    table.uuid('id').primary();
+    table.string('title').notNullable();
+    table.text('description').nullable();
+    table.string('severity', 8).notNullable();
```
Comment (mareklibra) [inline plugins/notifications-backend/migrations/20240313_broadcast.js:22]: Might become numeric within https://github.com/backstage/backstage/pull/23371.

```table.tinyint('severity').notNullable()```

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts [hunk 2]
```diff
@@ -94,13 +114,44 @@ export class DatabaseNotificationsStore implements NotificationsStore {
...
+  private mapBroadcastToDbRow = (notification: Notification) => {
+    return {
+      id: notification.id,
+      origin: notification.origin,
+      created: notification.created,
+      topic: notification.payload?.topic,
+      link: notification.payload?.link,
+      title: notification.payload?.title,
+      description: notification.payload?.description,
+      severity: notification.payload?.severity,
```
Comment (mareklibra) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:126]: recently considered to become numeric

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts [hunk 3]
```diff
@@ -229,51 +300,105 @@ export class DatabaseNotificationsStore implements NotificationsStore {
...
+            broadcasts.map(b => ({
+              broadcast_id: b.id,
+              user,
+              read,
+              saved,
+            })),
+          )
+          .onConflict(['broadcast_id', 'user'])
+          .merge(['read', 'saved']);
+        return;
```
Comment (mareklibra) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:371]: no need for the `return`

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts [hunk 4]
```diff
@@ -229,51 +300,105 @@ export class DatabaseNotificationsStore implements NotificationsStore {
...
+        await this.db('broadcast_user_status')
+          .insert(
+            broadcasts.map(b => ({
+              broadcast_id: b.id,
+              user,
+              read,
+              saved,
+            })),
+          )
+          .onConflict(['broadcast_id', 'user'])
```
Comment (mareklibra) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:369]: Is there a need to have two completely different flows per engine?

The 2nd one for the SQLite is not so nice but works and leaves us with less code to maintain.

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts [hunk 5]
```diff
@@ -30,9 +30,29 @@ const migrationsDir = resolvePackagePath(
   'migrations',
 );
 
+const NOTIFICATION_COLUMNS = [
```
Comment (mareklibra) [inline plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:33]: good way to go 👍

File: plugins/notifications-backend/migrations/20240313_broadcast.js (hunk 1 above)
Comment (drodil) [reply plugins/notifications-backend/migrations/20240313_broadcast.js:22]: Yeah, lots of conflicts incoming.

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts (hunk 4 above)
Comment (drodil) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:369]: I think sqlite is the exception here as this doesn't support this. This way is much faster and the proper way to do this.

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts (hunk 3 above)
Comment (drodil) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:371]: Will fix!

File: plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts (hunk 2 above)
Comment (drodil) [reply plugins/notifications-backend/src/database/DatabaseNotificationsStore.ts:126]: It's under discussion I think 🤔

**Review by mareklibra — APPROVED** [review]

**Review by mareklibra — APPROVED** [review]

**Pull request merged**